testeng/jobs/backupJenkins.groovy

```diff
@@ -0,0 +1,95 @@
+package testeng
+
+import org.yaml.snakeyaml.Yaml
+
+Map config = [:]
+Binding bindings = getBinding()
+config.putAll(bindings.getVariables())
+PrintStream out = config['out']
+
+try {
+    out.println('Parsing secret YAML file')
+    /* Parse k:v pairs from the secret file referenced by secretFileVariable */
+    String contents = new File("${BACKUP_JENKINS_SECRET}").text
+    Yaml yaml = new Yaml()
+    secretMap = yaml.load(contents)
+    out.println('Successfully parsed secret YAML file')
+}
+catch (any) {
+    out.println('Jenkins DSL: Error parsing secret YAML file')
+    out.println('Exiting with error code 1')
+    return 1
+}
+
+/*
+Example secret YAML file used by this script
+Config:
+    jenkinsInstance : test
+    volumeId : vol-123
+    region : us-west.1
+*/
+
+/* Iterate over the job configurations */
+secretMap.each { jobConfigs ->
+
+    Map jobConfig = jobConfigs.getValue()
+
+    assert jobConfig.containsKey('jenkinsInstance')
+    assert jobConfig.containsKey('volumeId')
+    assert jobConfig.containsKey('region')
+
+    job("backup-${jobConfig['jenkinsInstance']}-jenkins") {
+        
+        // private job
+        authorization {
+            blocksInheritance(true)
+            permissionAll('edx')
+        }
+
+        description('A regularly run job for creating snapshots of build jenkins master ' +
+                    ' and uploading them to Amazon s3')
+        // Keep logs longer than normal jenkins jobs
+        logRotator {
+            numToKeep(50)
+        }
+        concurrentBuild(false)
+        label("micro-worker")
+
+        // Configure the Exclusive Execution plugin, to reduce the amount of things in memory during snapshotting
+        configure { project ->
+            project / buildWrappers << 'hudson.plugins.execution.exclusive.ExclusiveBuildWrapper' {
+                skipWaitOnRunningJobs false
+            }
+        }
+
+        wrappers {
+            timeout {
+                absolute(20)
```

Comment:
First, about the data: I backed up Test Jenkins, which is about 9% the size of Build Jenkins, and it took about 5 minutes to back up. However, once the snapshot command has been issued, the snapshot process proceeds asynchronously, with regard to the Jenkins process. This timeout value will only come into play if the pip command is hanging. I suppose we should shorten it. 

However, this brings up a good question. What happens if the something goes wrong with aws. The script outputs some JSON data describing it's status. We can use Jenkins to verify something in the build process itself to determine if it should be considered a success. Then.. there is the issue of a snapshot starting, and then dying later on. I am not sure if it would make sense to make another call after some time to see if the snapshot no longer is marked as "pending"